BUG: inconsistent replace

QuentinN42 opened the pull request:
- [x] tests added
- [x] tests passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] release note in `doc/source/whatsnew/v1.1.3.rst`

Must fix some replace in IntBlock if called with float integer (such as 1.0).


closes #35376

QuentinN42 committed:
BUG: is_float added to IntBlock._can_hold_element

Added return is_integer(element) or is_float(element) to the IntBlock._can_hold_element method because an Block of ints can be replaced from int.
Read #35376 for more info

QuentinN42 committed:
BUG: is_float(e) and e.is_integer() added to IntBlock._can_hold_element

As @jbrockmendel said in #35376, you can replace an int by a float in an IntBlock only if the element is an integer.

QuentinN42 committed:
isort passed

move an import 2 lines above

simonjayhawkins commented:
Thanks @QuentinN42 for the PR. needs tests see https://github.com/pandas-dev/pandas/issues/35376#issuecomment-694733759 and a release note in doc/source/whatsnew/v1.1.3.rst

jreback commented on pandas/core/internals/blocks.py:
```diff
@@ -2066,7 +2067,7 @@ def _can_hold_element(self, element: Any) -> bool:
                 and self.dtype.itemsize >= tipo.itemsize
             )
-        return is_integer(element)
+        return is_integer(element) or (is_float(element) and element.is_integer())
```
this would only hit a non-inferred path, i don't think your added code does anything.

QuentinN42 replied:
If the element is a simple python float, `maybe_infer_dtype_type` return `None` so we are in this case.

jreback reviewed (changes requested):
pls always add a test that reproduces the issue *first* to make sure it fails.

QuentinN42 committed:
replace test added

Added the #35376 error as a test.

QuentinN42 commented:
> pls always add a test that reproduces the issue _first_ to make sure it fails.

Test added.

QuentinN42 committed:
release note

Added regression description in `doc/source/whatsnew/v1.1.3.rst`

QuentinN42 committed:
black on test_replace

moved 4 lines in once

simonjayhawkins commented on pandas/tests/frame/methods/test_replace.py:
```diff
@@ -974,6 +974,8 @@ def test_replace_for_new_dtypes(self, datetime_frame):
                 ),
             ),
+            # GH 35376
+            (DataFrame([[1, 1.0], [2, 2.0]]), 1.0, 5, DataFrame([[5, 5.0], [2, 2.0]]),),
```
can you add the .replace(1, 5) case and maybe also .replace(1.0, 5.0) and .replace(1, 5.0)

simonjayhawkins commented on doc/source/whatsnew/v1.1.3.rst:
```diff
@@ -34,7 +34,7 @@ Fixed regressions
 - Fixed regression in :meth:`Series.__getitem__` incorrectly raising when the input was a frozenset (:issue:`35747`)
 - Fixed regression in :meth:`read_excel` with ``engine="odf"`` caused ``UnboundLocalError`` in some cases where cells had nested child nodes (:issue:`36122`,:issue:`35802`)
--
+- Fixed regression in :meth:`DataFrame.replace` inconsistant replace when using a float in the replace method (:issue:`35376`)
```
```suggestion
- Fixed regression in :meth:`DataFrame.replace` inconsistent replace when using a float in the replace method (:issue:`35376`)
```

QuentinN42 committed:
Update doc/source/whatsnew/v1.1.3.rst

Thx @simonjayhawkins for the typo

Co-authored-by: Simon Hawkins <simonjayhawkins@gmail.com>

dsaxton commented:
@QuentinN42 Can you merge master to fix merge conflict? Also I think this doesn't need to be a draft.

QuentinN42 commented:
Wait...
I think #34878 close #36444...

QuentinN42 committed:
Merge branch 'master' into issue35376-inconsistant-replace

# Conflicts:
#	doc/source/whatsnew/v1.1.3.rst

QuentinN42 committed:
more tests on test_replace

added .replace(1, 5), .replace(1.0, 5.0) and .replace(1, 5.0)

pep8speaks commented:
Hello @QuentinN42! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/tests/frame/methods/test_replace.py`](https://github.com/pandas-dev/pandas/blob/473f1083b31ca69522ff45e521a151783a70f967/pandas/tests/frame/methods/test_replace.py):

> [Line 980:89](https://github.com/pandas-dev/pandas/blob/473f1083b31ca69522ff45e521a151783a70f967/pandas/tests/frame/methods/test_replace.py#L980): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 88 characters)

QuentinN42 committed:
test_replace PEP8

Moved all singles lines in 6 lines

QuentinN42 commented:
I've rolled back on typo from f64817b to pass the PEP8 test.
But it may not pass the CI / Checks (PR)

QuentinN42 committed:
black on test_replace

Formatting is just non sense

QuentinN42 commented:
How can we see wich tests have failed and what is the given output ?

simonjayhawkins commented:
> How can we see wich tests have failed and what is the given output ?

may be fixed by #36511

QuentinN42 commented:
> > How can we see wich tests have failed and what is the given output ?
> 
> may be fixed by #36511

 #36511 has been merged.

simonjayhawkins commented:
> #36511 has been merged.

i've restarted the tests since (azure normally picks up the changes in master) but it looks like there are still failures.

can you merge upstream/master into this branch

QuentinN42 committed:
Merge remote-tracking branch 'upstream/master' into issue35376-inconsistant-replace

jreback commented on pandas/core/internals/blocks.py:
```diff
@@ -2066,7 +2067,7 @@ def _can_hold_element(self, element: Any) -> bool:
                 and self.dtype.itemsize >= tipo.itemsize
             )
-        return is_integer(element)
+        return is_integer(element) or (is_float(element) and element.is_integer())
```
@jbrockmendel are you sure you think the patch should be *here*; this is hit by a lot of paths where this doesn't make any sense. we almost always want to take some sort of action based on return value here if it fails. e.g. why is this not more specific to replace?

dsaxton replied:
Also is this simply wrong:
```python
[ins] In [3]: is_integer(1.0)
Out[3]: False
```
especially since
```python
[ins] In [4]: 1.0.is_integer()
Out[4]: True
```

jreback replied:
ahh its doing an equality check under the hood, we are not

```
n [60]: 1.0.is_integer()                                                                                                                                  
Out[60]: True

In [61]: 1.1.is_integer()                                                                                                                                  
Out[61]: False
```

jreback replied:
maybe could replace this entirely with `element.is_integer()` and call it a day

jreback replied:
though my concern is *still* that the purpose of this check is *can* we hold an integer, not are we a float and *can* be treated as an integer. though practically it may not make a difference.

QuentinN42 replied:
So I don't realy understand ...
Need I replace `return is_integer(element)` into `return element.is_integer()`.
If I do that, need I try except the `AttributeError` and return `None` ?

QuentinN42 replied:
Or `False` ?

jbrockmendel replied:
Longer-term we're going to have to look more carefully about can_hold_element and what we want it to mean.  In the short run, for this check to be consistent with the tipo check above it should exclude floats

(so my previous suggestion may have been unhelpful, sorry)

jreback replied:
ok this is fine @QuentinN42 if you can add a comment here would be good

jreback reviewed (changes requested):
pls ping on green

QuentinN42 committed:
added comment in IntBlock._can_hold_element

marking GH 36444 and GH 35376 close to (is_float(element) and element.is_integer())

QuentinN42 commented:
I have a feeling that there are reproducibility errors with black.
On my personals repos, I've removed it from pre-commit for this errors.

On server :
```
black....................................................................Failed
- hook id: black
- files were modified by this hook

reformatted /home/runner/work/pandas/pandas/pandas/tests/frame/methods/test_replace.py
All done! ✨ 🍰 ✨
1 file reformatted, 1133 files left unchanged.

flake8...................................................................Passed
flake8-pyx...............................................................Passed
flake8-pxd...............................................................Passed
isort....................................................................Passed
pyupgrade................................................................Passed
```
On local :
```
black....................................................................Passed
flake8...................................................................Failed
- hook id: flake8
- exit code: 1

pandas/core/generic.py:1:1: F407 future feature annotations is not defined
pandas/core/algorithms.py:5:1: F407 future feature annotations is not defined
pandas/core/construction.py:7:1: F407 future feature annotations is not defined
pandas/core/frame.py:11:1: F407 future feature annotations is not defined

flake8-pyx...............................................................Passed
flake8-pxd...............................................................Passed
isort....................................................................Passed
```

The error come from this line (`pandas/tests/frame/methods/test_replace.py`):
```
(DataFrame([[1, 1.0], [2, 2.0]]), 1, 5, DataFrame([[5, 5.0], [2, 2.0]]),),
```
that black want to split into :
```
(
    DataFrame([[1, 1.0], [2, 2.0]]),
    1,
    5,
    DataFrame([[5, 5.0], [2, 2.0]]),
),
```

simonjayhawkins commented:
@QuentinN42 the version of black used on ci was changed yesterday. see #36493

updating black locally and merging upstream/master will probably fix.

QuentinN42 committed:
Merge remote-tracking branch 'upstream/master' into issue35376-inconsistant-replace

QuentinN42 committed:
black test_replace.py

refactor

QuentinN42 commented:
The error come from `pandas/tests/test_sorting.py::TestSafeSort::test_codes_out_of_bound`.
Not me :cry: (I hope).

simonjayhawkins commented:
> The error come from `pandas/tests/test_sorting.py::TestSafeSort::test_codes_out_of_bound`.
> Not me 😢 (I hope).

seen b4 https://github.com/pandas-dev/pandas/pull/36382#issuecomment-692707377

restarted ci

QuentinN42 committed:
Merge remote-tracking branch 'upstream/master' into issue35376-inconsistant-replace

# Conflicts:
#	doc/source/whatsnew/v1.1.3.rst

simonjayhawkins commented:
@jreback green https://github.com/pandas-dev/pandas/pull/36444#pullrequestreview-493882423

simonjayhawkins reviewed (approved):
Thanks @QuentinN42 lgtm

jreback commented on pandas/core/internals/blocks.py:
```diff
@@ -2066,7 +2067,8 @@ def _can_hold_element(self, element: Any) -> bool:
                 and self.dtype.itemsize >= tipo.itemsize
             )
-        return is_integer(element)
+        # Will have to be modified in the future see GH 36444 and GH 35376
```
not what I meant.
say something like
```
we have not inferred an integer; check if we have an equal float
```

jreback commented:
the current comment is not useful - needs to be updated

QuentinN42 committed:
changing comment to be more explicit

QuentinN42 commented:
I've changed into :
```
We have not inferred an integer from the dtype
check if we have a builtin int or a float equal to an int
```

QuentinN42 committed:
Merge remote-tracking branch 'upstream/master' into issue35376-inconsistant-replace

simonjayhawkins committed:
Merge remote-tracking branch 'upstream/master' into issue35376-inconsistant-replace

simonjayhawkins commented:
@jreback ok to merge on green

jreback commented:
thanks @QuentinN42

simonjayhawkins commented:
@meeseeksdev backport 1.1.x
